tests/postgres_tests/test_operations.py

```diff
@@ -411,3 +412,57 @@ def test_remove(self):
         self.assertEqual(name, 'RemoveCollation')
         self.assertEqual(args, [])
         self.assertEqual(kwargs, {'name': 'C_test', 'locale': 'C'})
+
+
+@unittest.skipUnless(connection.vendor == 'postgresql', 'PostgreSQL specific tests.')
+@modify_settings(INSTALLED_APPS={'append': 'migrations'})
+class AddConstraintNotValidTests(OperationTestBase):
+    app_label = 'test_add_constraint_not_valid'
+
+    def test_add(self):
+        project_state = self.set_up_test_model(self.app_label)
+        table_name = '%s_pony' % self.app_label
+        constraint = CheckConstraint(check=Q(pink__gte=4), name='pony_pink_check')
+        new_state = project_state.clone()
+        operation = AddConstraintNotValid(model_name='pony', constraint=constraint)
+        self.assertEqual(
+            operation.describe(),
+            'Create constraint pony_pink_check on model pony without enforcing validation on existing rows'
+        )
+        self.assertEqual(operation.migration_name_fragment, "pony_pony_pink_check_not_valid")
+        operation.state_forwards(self.app_label, new_state)
+        self.assertEqual(len(new_state.models[self.app_label, 'pony'].options['constraints']), 1)
+        self.assertConstraintNotExists(table_name, 'pony_pink_check')
+        # Add constraint.
+        with connection.schema_editor(atomic=True) as editor:
+            operation.database_forwards(self.app_label, editor, project_state, new_state)
+        self.assertConstraintExists(table_name, 'pony_pink_check')
+        # Reversal.
+        with connection.schema_editor(atomic=True) as editor:
+            operation.database_backwards(self.app_label, editor, project_state, new_state)
+        self.assertConstraintNotExists(table_name, ['pony_pink_check'])
+
+
+@unittest.skipUnless(connection.vendor == 'postgresql', 'PostgreSQL specific tests.')
+@modify_settings(INSTALLED_APPS={'append': 'migrations'})
+class ValidateConstraintTests(OperationTestBase):
+    app_label = "test_validate_constraint"
+
+    def test_validate(self):
+        project_state = self.set_up_test_model(self.app_label)
+        table_name = "%s_pony" % self.app_label
+        with connection.schema_editor(atomic=True) as editor:
+            editor.execute('INSERT INTO %s VALUES(default, -1, 1.2)' % editor.quote_name(table_name))
+            editor.execute('ALTER TABLE %s ADD CONSTRAINT %s CHECK (pink>0) NOT VALID' %
+                           (editor.quote_name(table_name), "pony_pink_check")
+                           )
```

Comment:
Why we're adding a constraint manually and not with `AddConstraintNotValid()`?

Also, please use hanging indentation.

Reply:
I didn't think it really matters if we add the constraint via plain SQL or use the operation. Is there any specific reason why we would prefer using the operation in this case?

Reply:
> Is there any specific reason why we would prefer using the operation in this case?

Yes, because we have it. Using a RAW SQL is the last option, we're developing the ORM in order not to use them.